software/firmware/europi_script.py

```diff
@@ -52,23 +52,111 @@ def display_name(cls):
         """
         return cls.__qualname__
 
-    #
-    # private functions for now, with the intention that they are eventually added to the public API
-    #
+    # Save State Methods
+
+    @property
+    def _state_filename(self):
+        return f"saved_state_{self.__class__.__qualname__}.txt"
+
+    def save_state(self, state: str):
+        """Take state in persistence format as a string and write to disk."""
+        return self._save_state(state)
+
+    def save_state_bytes(self, state: bytes):
+        """Take state in persistence format as bytes and write to disk.
+
+        .. note::
+            Be mindful of how often `save_state_bytes()` is called because
+            writing to disk too often can slow down the performance of your
+            script. Only call save state when state has changed and consider
+            adding a time since last save check to reduce save frequency.
+
+        ::
+
+            # https://docs.python.org/3/library/struct.html#format-characters
```

Comment:
The struct stuff is nice, but I wonder if it unnecessarily complicates the example.

Perhaps there should just be a minimal example, followed by a mention that a struct could be useful, and the link.